sympy/simplify/cse_main.py

```diff
@@ -136,6 +137,238 @@ def postprocess_for_cse(expr, optimizations):
     return expr
 
 
+class FuncArgTracker(object):
+    """
+    A class which manages a mapping from functions to arguments and an inverse
+    mapping from arguments to functions.
+    """
+
+    def __init__(self, funcs):
+        # To minimize the number of symbolic comparisons, all function arguments
+        # get assigned a value number.
```

Comment:
Although I tend to *think* that this matters, I thought that a hash code or something like that is used to tell if two items are the same or not. But a little timing indicates that it matters a LOT:

```
>>> timeit('1==3')
0.052123287547175275
>>> timeit('x==y','from sympy import S;from sympy.abc import x,y;x=S(1);y=S(2)')
1.5453416996564044
>>> timeit('x==y','from sympy.abc import x,y')
12.901627279220296
```